File: src/python/api/quiz.py
from flask import Blueprint, request, jsonify

from src.python.engine.db import Db
from ..auth import auth

api_quiz = Blueprint("quiz", __name__, url_prefix="/api/quiz")


@api_quiz.route("/", methods=["POST"])
@auth.login_required
def r_quiz_build():
    r = request.json
    return jsonify({"cards": [dict(c) for c in Db.parse_cond(r["cond"], fields=[
        "_id", "entry", "type"
    ])]})


@api_quiz.route("/render", methods=["POST"])
@auth.login_required
def r_quiz_render():
    r = request.json

    if r.get("cardId"):
        return jsonify(Db.render_from_id(card_id=r["cardId"]))
    else:
        return jsonify(Db.render(r["entry"], r["type"]))


@api_quiz.route("/right", methods=["PUT"])
@auth.login_required
def r_quiz_right():
    r = request.json
    Db.mark_right(r.get("id"), r.get("data"))
    return jsonify({"error": None})


@api_quiz.route("/wrong", methods=["PUT"])
@auth.login_required
def r_quiz_wrong():
    r = request.json
    Db.mark_wrong(r.get("id"), r.get("data"))
    return jsonify({"error": None})

File: src/node/api/quiz.ts
import { Router } from "express";
import asyncHandler from "express-async-handler";
import needUserId from "../middleware/needUserId";
import auth from "../auth/token";
import Database from "../engine/db";

const router = Router();
router.use(auth.optional);
router.use(needUserId());

router.post("/", asyncHandler(async (req, res) => {
    const {cond} = req.body;
    const db = new Database();
    const cards = await db.parseCond(res.locals.userId, cond, {
        fields: ["_id", "entry", "type"]
    });

    return res.json(cards);
}));

router.post("/render", asyncHandler(async (req, res) => {
    const {id, entry, type} = req.body;
    const db = new Database();
    if (id) {
        return res.json(await db.renderFromId(res.locals.userId, id));
    } else {
        return res.json(await db.render(entry, type));
    }
}));

router.put("/right", asyncHandler(async (req, res) => {
    const {id, data} = req.body;
    const db = new Database();
    return res.json({
        id: await db.markRight(res.locals.userId, id, data)
    });
}));

router.put("/wrong", asyncHandler(async (req, res) => {
    const {id, data} = req.body;
    const db = new Database();
    return res.json({
        id: await db.markWrong(res.locals.userId, id, data)
    });
}));

export default router;
